Support Auto Update

thatStrangeGuyThorchain opened the pull request:
to test locally play with version at the `package.json` (e.g. "version": "0.0.9")

NOTE:  update-info requests for every application start

closes #889

thatStrangeGuyThorchain committed:
test autoupdate

thatStrangeGuyThorchain committed:
tmp

thatStrangeGuyThorchain committed:
added console.log

thatStrangeGuyThorchain committed:
changed version

thatStrangeGuyThorchain committed:
log info and new version

thatStrangeGuyThorchain committed:
log info and new version

thatStrangeGuyThorchain committed:
test

thatStrangeGuyThorchain committed:
test

thatStrangeGuyThorchain committed:
tmp

thatStrangeGuyThorchain committed:
v7

thatStrangeGuyThorchain committed:
v8

thatStrangeGuyThorchain committed:
v9

thatStrangeGuyThorchain committed:
v10

thatStrangeGuyThorchain committed:
v11

thatStrangeGuyThorchain committed:
v12

thatStrangeGuyThorchain committed:
clean up

thatStrangeGuyThorchain committed:
tmp deisabled another OSs

thatStrangeGuyThorchain committed:
clean up

thatStrangeGuyThorchain committed:
fixed translations

thatStrangeGuyThorchain committed:
v14

thatStrangeGuyThorchain committed:
v15

thatStrangeGuyThorchain committed:
v16

thatStrangeGuyThorchain committed:
added a way to mock update info while developing

thatStrangeGuyThorchain committed:
fixed i18n; changed goToUpdates cb

thatStrangeGuyThorchain committed:
v17

thatStrangeGuyThorchain committed:
fixed package.json

thatStrangeGuyThorchain committed:
reverted menu/view

thatStrangeGuyThorchain committed:
added missed mock files

veado commented on src/renderer/services/appUpdate/service.ts:
```diff
@@ -0,0 +1,20 @@
+import * as FP from 'fp-ts/function'
+import * as O from 'fp-ts/Option'
+import * as Rx from 'rxjs'
+import * as RxOp from 'rxjs/operators'
+
+import { observableState } from '../../helpers/stateHelper'
+
+type AppUpdater = O.Option<string>
+
+const { get$: appUpdater$, set: setAppUpdater } = observableState<AppUpdater>(O.none)
+
+FP.pipe(
+  Rx.from(window.apiAppUpdate),
+  RxOp.catchError(() => Rx.EMPTY),
+  RxOp.map((version) => O.some(version))
+).subscribe(setAppUpdater)
+
+const resetAppUpdater = () => setAppUpdater(O.none)
+
+export { appUpdater$, resetAppUpdater }
```
Put all of this logic of `appUpdate/service.ts ` into a hook `useAppUpdate`

veado commented on src/renderer/services/appUpdate/service.ts:
```diff
@@ -0,0 +1,20 @@
+
+FP.pipe(
+  Rx.from(window.apiAppUpdate),
+  RxOp.catchError(() => Rx.EMPTY),
```
Handle error (we do need to know if something went wrong)

thatStrangeGuyThorchain replied:
ok. in which way ? show something to the user or just internal `console.log` ?

veado replied:
Show same modal "New version is available", but with an error message (in red) in it's body: "`Check for update failed: ${error.?message ?? error.toString()}`" or similar

veado commented on src/renderer/i18n/de/update.ts:
```diff
@@ -0,0 +1,9 @@
+import { UpdateMessages } from '../types'
+
+const update: UpdateMessages = {
+  'update.available': 'New version is available - DE',
+  'update.description':
+    'There is a new version {version} available. You can go to the releases page and download it. - DE'
+}
```
```
'update.available': 'Neues Update verfügbar',
'update.description':
    'Eine neue Version {version} ist verfügbar. Bei Klick auf "Ok" gelangst Du zur Release Webseite, wo Du die neue Version downloaden kannst.'
```

veado commented on src/renderer/i18n/en/common.ts:
```diff
@@ -78,7 +78,8 @@ const common: CommonMessages = {
   'common.theme.light': 'Day mode',
-  'common.theme.dark': 'Night mode'
+  'common.theme.dark': 'Night mode',
+  'common.go': 'Go'
```
```'common.ok' ='Ok'```

veado commented on src/renderer/i18n/de/common.ts:
```diff
@@ -78,7 +78,8 @@ const common: CommonMessages = {
   'common.theme.light': 'Tag modus',
-  'common.theme.dark': 'Nacht modus'
+  'common.theme.dark': 'Nacht modus',
+  'common.go': 'Go'
```
'common.ok' ='Ok'

veado commented on src/main/preload.ts:
```diff
@@ -27,3 +29,7 @@ contextBridge.exposeInMainWorld('apiHDWallet', apiHDWallet)
+
+contextBridge.exposeInMainWorld('apiAppUpdate', getApiAppUpdate())
+
+ipcRenderer.on(IPCMessages.LOG, (...args) => console.log('--- log ---', ...args))
```
Still needed: `console.log...` ?

thatStrangeGuyThorchain replied:
nope

veado commented on src/renderer/i18n/en/update.ts:
```diff
@@ -0,0 +1,8 @@
+import { UpdateMessages } from '../types'
+
+const update: UpdateMessages = {
+  'update.available': 'New version is available',
+  'update.description': 'There is a new version {version} available. You can go to the releases page and download it'
```
```
'update.available': 'New update available',
'update.description':
    'There is a new version {version} available. Click "Ok" to go to the release page, where you can download it.'
```

thatStrangeGuyThorchain committed:
added loading info directly from GH

thatStrangeGuyThorchain committed:
changed i18n; clean up

thatStrangeGuyThorchain committed:
reverted version

thatStrangeGuyThorchain committed:
removed dev-app-update

thatStrangeGuyThorchain committed:
added description to readme about how to check auto-update

veado commented on src/main/electron.ts:
```diff
@@ -140,15 +141,44 @@ const initIPC = () => {
+  await autoUpdater.checkForUpdates()
 }
 
+autoUpdater.on('update-available', (info: { version: string }) => {
+  sendUpdateMessage(IPCMessages.UPDATE_AVAILABLE, info.version)
+})
+
+autoUpdater.on('update-not-available', () => {
+  sendUpdateMessage(IPCMessages.UPDATE_NOT_AVAILABLE)
+})
```
Move it into `main/api/appUpdate.ts `

veado commented on src/main/electron.ts:
```diff
@@ -140,15 +141,44 @@ const initIPC = () => {
   })
 }
 
+autoUpdater.logger = {
+  ...log,
+  // Set custom info logger for better understanding in common debug-flow
+  info(message?: unknown): void {
+    console.log('[AutoUpdater.info]', message)
+  }
+}
+interface SendUpdateMessage {
+  (message: IPCMessages.UPDATE_AVAILABLE, version: string): void
+  (message: IPCMessages.UPDATE_NOT_AVAILABLE): void
+}
+const sendUpdateMessage: SendUpdateMessage = (message: IPCMessages, version?: string) => {
+  if (mainWindow) {
+    mainWindow.webContents.send(message, version)
+  }
+}
+
+// Disable autoDownload
+autoUpdater.autoDownload = false
+autoUpdater.allowPrerelease = true
```
Move it into `main/api/appUpdate.ts `

veado commented on src/main/electron.ts:
```diff
@@ -140,15 +141,44 @@ const initIPC = () => {
   app.on('activate', activateHandler)
-
   initIPC()
+  await autoUpdater.checkForUpdates()
```
Extract `autoUpdater.checkForUpdates()` into `main/api/appUpdate.ts` to manage requests for checking updates from `renderer` side something like this:
```
// renderer -> main
CHECK_APP_UPDATE

// renderer -> main
APP_UPDATE
```

With this, we can start requests every 6 (or 12hrs etc.) to let users know about important updates, even if an user has open its wallet for two days or so...

veado commented on package.json:
```diff
@@ -1,14 +1,14 @@
     "dev": "concurrently --kill-others --success first \"yarn:dev:*\"",
     "dev:app": "cross-env BROWSER=none yarn start:web",
-    "dev:electron": "wait-on http://localhost:3000 && nodemon -I --watch ./src/main/ --exec electron ./public/dev.js",
+    "dev:electron": "wait-on http://localhost:3000 && nodemon -I --watch ./src/main/ --exec electron .",
```
`--exec electron .` -> `--exec electron ./index.js` (just to make it more clear for non JS users)

thatStrangeGuyThorchain replied:
unfortunately we can not do that for the [same issue](https://github.com/electron/electron/issues/7085#issuecomment-354305046) we discussed with you.
P.S. yep. i've tried to use `./index.js` directly and there is comparison with `electro`'s version
![image](https://user-images.githubusercontent.com/67793194/116727472-48e14c00-a9ed-11eb-82bc-ce02a085b167.png)

thatStrangeGuyThorchain committed:
fixed type at readme

thatStrangeGuyThorchain committed:
refactored check for updates to be triggered every 6 hours; moved logic to the useAppUpdate from services/appUpdate

thatStrangeGuyThorchain committed:
refactored AppUpdateModal to be an Alert component

thatStrangeGuyThorchain committed:
updated i18n

thatStrangeGuyThorchain committed:
added ExternalLink icon to the ok button

thatStrangeGuyThorchain committed:
added error handling for AppUpdate

thatStrangeGuyThorchain committed:
added stories for AppUpdate

thatStrangeGuyThorchain committed:
reverted package version

thatStrangeGuyThorchain committed:
clean up

veado commented on src/renderer/views/app/AppView.tsx:
```diff
@@ -57,7 +61,10 @@ export const AppView: React.FC = (): JSX.Element => {
   return (
     <Styled.AppWrapper>
       <Styled.AppLayout>
+        {isDesktopView && <AppUpdateView />}
         <Header />
+        {!isDesktopView && <AppUpdateView />}
```
```ts
<AppUpdateView />
<Header />
...
```
No check for `isDesktopView` needed. _I know we discussed the other way, but this looks better in UI._

veado commented on src/renderer/i18n/de/update.ts:
```diff
@@ -0,0 +1,9 @@
+const update: UpdateMessages = {
+  'update.description': 'Neue Version {version} verfügbar',
+  'update.link': 'Release Webseite',
+  'update.checkFailed': 'Error while check for update: {error} - DE'
```
```
Error beim Update-Check: {error}
```

veado commented on src/main/api/appUpdate.ts:
```diff
@@ -0,0 +1,60 @@
+  ...log,
+  // Set custom info logger for better understanding in common debug-flow
+  info(message?: unknown): void {
+    console.log('[AutoUpdater.info]', message)
```
^ In `dev` mode only?

thatStrangeGuyThorchain committed:
autoupdate logger settings only for dev

thatStrangeGuyThorchain committed:
changed i18n

thatStrangeGuyThorchain committed:
text v0.1.1

thatStrangeGuyThorchain committed:
moved autoUpdater settings inside of registerAppCheckUpdatedHandler

thatStrangeGuyThorchain committed:
refactored AppView to render AppUpdateView only before Header; refactored Header propup to be rendered right after Header container component

thatStrangeGuyThorchain committed:
refactored styles for AppUpdate component

thatStrangeGuyThorchain committed:
added comments for using custom calculation of getHeaderBottomPosition

thatStrangeGuyThorchain committed:
reverted version number

veado commented on dev-app-update.sample.yml:
```diff
@@ -0,0 +1,5 @@
+owner: 'thorchain'
+repo: 'asgardex-electron'
+# @see https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token
+token: githubPersonalAccessToken
```
[nitpick]
```
# GitHub personal access token
# @see https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token
token: XXX
```

veado commented on src/main/api/appUpdate.ts:
```diff
@@ -0,0 +1,62 @@
+        })
+
+        autoUpdater.once('update-not-available', () => {
+          resolve(RD.initial)
```
It should be still `success` to handle a "Not available" status at `render` side. Also we do need to handle `pending` state (here in `main` or in `renderer` thread). Because as a next step we will let the user run a `re-check` (e.g. in case of `failure` or by providing a "check for updates" button in `Settings`). Not needed with this PR, just to let you know about it.

thatStrangeGuyThorchain replied:
yeah. i thought about `RD.success` but in this case we have to have `RD.RemoteData<Error, O.Option<string>>` to understand if there is "real" update available which i thought a little bit redundant. if this is just an `RD.RemoteData<Error, string>` we can just resolve Promise with `RD.initial` state to tell to the renderer process that there is no real update available. does it make sense? if no - i can re-factor with `RemoteData<Error, Option<string>>`

about `RD.pending`. it should be handled at the `useAppUpdate`-level (when requesting/re-requesting info about updates) as as  soon as we use promise (native way for electron) to communicate between `main` and `renderer` processes we can not send more than a single value so from `main` process we can send **only** a single value

veado commented on src/main/ipc/messages.ts:
```diff
@@ -8,6 +8,9 @@ enum IPCMessages {
   GET_LEDGER_ADDRESS = 'GET_LEDGER_ADDRESS',
   SEND_LEDGER_TX = 'SEND_LEDGER_TX',
+  UPDATE_AVAILABLE = 'UPDATE_AVAILABLE',
+  UPDATE_NOT_AVAILABLE = 'UPDATE_NOT_AVAILABLE',
```
Are both messages ^ still needed?

veado commented on src/renderer/components/AppUpdate/AppUpdate.tsx:
```diff
@@ -0,0 +1,75 @@
+  const intl = useIntl()
+  const isDesktopView = Grid.useBreakpoint()?.lg ?? false
+
+  if (props.isOpen && props.type === 'success') {
```
Pls destruct values from `props` before (as we do it in other React components).

veado commented on src/renderer/components/AppUpdate/AppUpdate.tsx:
```diff
@@ -0,0 +1,75 @@
+import * as Styled from './AppUpdate.styles'
+
+export type AppUpdateModalProps =
+  | {
+      isOpen: true
+      type: 'success'
+      goToUpdates: () => void
+      version: string
+      close: () => void
+    }
+  | {
+      isOpen: true
+      type: 'fail'
+      close: () => void
+      message: string
+    }
+  | {
+      isOpen: false
+    }
```
Why do we need union types for props here ^?

veado commented on src/renderer/views/app/AppUpdateView.tsx:
```diff
@@ -0,0 +1,33 @@
+          (version): AppUpdateModalProps => ({
+            isOpen: true,
+            type: 'success',
+            goToUpdates: () => window.apiUrl.openExternal(`${packageInfo.repository.url}/releases/tag/v${version}`),
```
[nitpick] Let's use `packageInfo.repository.url` in `ExternalUrl.GITHUB` (`src/shared/const.ts`). Or remove `ExternalUrl.GITHUB` there to use `packageInfo.repository.url` everywhere. _Just to avoid duplication ..._

thatStrangeGuyThorchain committed:
replaced ExternalUrl.GITHUB_REPO with package.json's values across the project;
dev-app-update.sample.yml: changed description;
package.json: fixed bugs.url path

thatStrangeGuyThorchain committed:
src/main/ipc: remove UPDATE_NOT_AVAILABLE message

thatStrangeGuyThorchain committed:
removed usage of package.json file
